New Algorithm: Abins2D

ajjackson opened the pull request:
**Description of work.**

Add a new Abins2D algorithm. This shares a lot of code with the existing Abins algorithm, but provides a different front-end and returns 2-D outputs. These are suitable for a different set of instruments with different user parameters; as there are no immediate plans to develop a complex Abins Interface, this split over two algorithms seems the best bet for providing a reasonable user interface in each case.

There have been a few previous iterations of this PR. This one is based on a rework of the underlying maths to efficiently simulate spectra in 2-D q-energy space.

Several ISIS instruments are supported, making use of PyChop to compute resolution functions and identify angle limits. Rather than use pychop for every point, a regular set of points is computed and a polynomial fitted to obtain width (sigma) as a function of energy. This is then passed to the instrumental broadening algorithm which uses linear interpolation to reduce the number of Gaussian kernel evaluations.

Additionally, the ILL-PANTHER instrument is included. The resolution-E_i-Delta_E relationship is fitted to data provided by ILL.

**Todo:**
- [x] rename TwoDMap to Ideal2D
- [x] add systemtest
- [x] get doctests working

**To test:**
Unit tests, systemtest and doctests are provided.

I would also request a manual test if possible; open Abins2D, play with parameters, load a .phonon file. It should support the same input files as Abins.

**Release notes:**
This is a substantial, long-awaited new feature.
A release note has been added under Direct, as this algorithm is for simulation of direct-geometry instruments. Note that previous Abins notes have always been classified as "Indirect"; this raises questions about how best to classify Abins developments going forward.

- [x] Add a pretty picture to release note

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

ajjackson committed:
WIP: Abins2D prototype in E-Q space

ajjackson committed:
WIP: Efficient Abins2D: demo plot

ajjackson committed:
WIP: Efficient Abins2D - create workspace outputs

ajjackson committed:
Abins2D: Add ILL-PANTHER instrument

Resolution function based on fitting to data from ILL.

It looks rather broad, need to double check that's ok...
But the main issue here is if we use an unphysically large energy
range. Constraining incident energy to the known instrument range
should help manage this. The next step would be to overhaul energy/bin
handling in general; currently Abins constructs energy bins all over
the place by referring to limits in abins.parameters. This is not
flexible enough to handle these cases without a lot of redundancy;
instead we could consider having the Algorithm specify a set of bins
which are then passed around.

This might be a good time to also drop "bin_width" in favour of n_bins?

ajjackson committed:
Abins: Panther instrument; pychop resolution units; s array shapes

- Division by unit conversion was incorrect, giving results that were
  far too sharp. Polyfit is done in meV so need to convert _to_
  wavenumber here.
- Panther instrument file missing from previous commit!

ajjackson committed:
Abins: improve Panther resolution function

- Initial resolution function was based on backwards energy scale,
  giving too-broad results under most conditions.
- Incident energy dependence is made slightly more complex to improve
  fit quality.

ajjackson committed:
Abins 2D: Add MERLIN, refactor 2D instruments

Introduce a DirectInstrument class which handles e_init setting and
kinematic constraints; this reduces redundancy and keeps
instrument-specific classes more focused.

MERLIN is just another Pychop instrument, so very straightforward to
add. Some of these Pychop parameters are consistent actually, maybe
that redundancy can also be cut...

ajjackson committed:
Abins: refactor parameter access, especially energy ranges

- Add some utility methods for parameter lookup. These handle the
  logic of checking different parameter levels, which was repeated in
  various places.
  - This change is intended to cut redundancy and improve readability
- Instruments become responsible for knowing/reporting energy range
  - For direct instruments this is based on incident energy whereas
    for indirect this is an instrument parameter
  - Incident energy gains a public getter
  - Correct upper energy-transfer limit for Lagrange: max(E_init) - E_final
- Fix Abins2D bug reading instrument limits

One strange bug was noticed in testing: MAPS spectra are blank for A
chopper and low incident energies (< 100 meV).

ajjackson committed:
Abins2D: use updated Pychop API

- The PyChop2 interface is deprecated and asks that we use the
  Instrument interface.

ajjackson committed:
Abins: fix fine bins initialisation

Typo during the previous commit refactor - was leading to a useless
small set of fine bins and empty data.

ajjackson committed:
Abins: fix logic for interpolation between two kernels

This was not experienced when testing 1D examples where there was
always a decent range of broadening widths. However, when using a
Pychop instrument at low E_init it is possible that only two sigma
samples are needed for interpolation. In that case the existing logic
is broken, e.g.::

  n_kernels = ceil( log(1.3) / log(1.41) ) = 1

To interpolate properly we will need two kernels, one at min(sigma)
and one at min(sigma) * 1.41. But the n=1 case is caught by
``if n_kernels == 1:`` and we get a single kernel width [min(sigma)].

This is the correct outcome for a _constant_ width, in which case::

  n_kernels = ceil( log(min_sigma / min_sigma) / log(1.41) )
            = ceil(0.) = 0

To fix and clarify the logic, move the (+ 1) into n_kernels before the
if statement.

ajjackson committed:
Abins2D: enable MERLIN instrument in Algorithm interface

The instrument was already defined, just needed to add to the list.

ajjackson committed:
Abins: clean up comments and restore 1-D results caching

2-D caching is slow compared to calculation time, so remains disabled
until this can be improved.

ajjackson committed:
WIP Abins2D unit conversion

This implementation would be fairly clean, but doesn't work because
the ConvertUnits algorithm doesn't play well with simulated
workspaces. (It wants sample/instrument information that doesn't exist.)

ajjackson committed:
Abins: refactor algorithm inheritance, cut Abins2D redundancy

The main issue being solved here is the maintenance problem of needing
to assign local properties in any class derived from AbinsAlgorithm.

The opportunity is also taken to move "Autoconvolution" into the
common parameter set as the feature is available in both Abins and Abins2D.

There is an extensive use of class variables to store state, which
seems problematic; it would lead to strange behaviour if we try to run
multiple instances of Abins simultaneously.

I haven't eliminated all of those here, but taken the opportunity to
move things into __init__ statements while cutting the redundant elements.

ajjackson committed:
Abins2D: Convert output units

ConvertUnits isn't working for this, so we just adjust the bin values
when creating workspaced. It doesn't really matter that everything was
calculated in cm-1...

Well, actually there is a problem which is that the spectrum units are
not properly normalised. Spectral values should really be something
like "barn/energy" and so vary with energy units. But Mantid likes to
normalise spectra to counts/bin. As we haven't changed the size of the
bin or the incoming z-values, the spectrum value also stays the same.

The problem is that this makes the values sensitive to the bin size
used in calculation; reasonable in an experimental context but it
doesn't deliver the correct underlying property. Abins2D spectra already
have different intensity with different bin size... Abins(1D) doesn't
have that problem. Why is it different? I don't see an explicit correction.

ajjackson committed:
Abins2D: Kinematic constraints

Well, that came together more easily than expected! (The prep work was
done when computing the max/min q limits; I was worried that marking
empty elements with NaN would cause trouble somewhere but this seems fine.)

ajjackson committed:
Abins2D: clean up and fix TwoDMap instrument

- The TwoDMap instrument provides an "ideal" simulation with a fairly
  fine resolution function, large q range and no kinematic
  constraints.
- The q range and energy are determined by abins.parameters so can be
  changed on-the-fly if necessary, but the defaults should be suitable
  for most use cases.

ajjackson commented:
The tests are failing for an interesting reason and I am not entirely sure how to proceed.

I think the problem is that the Pychop-based instruments are importing a useful class from mantidqtinterfaces.PyChop . This resides in the qt part of the codebase; it seems this is forbidden and a test enforces the separation? Presumably this is intended to avoid problems when running mantidpython without Workbench.

The thing is that most of this code doesn't really need to be blocked off like this - including all the parts used by Abins. There is _qt/python/mantidqtinterfaces/mantidqtinterfaces/PyChop.py_, which loads a GUI from _qt/python/mantidqtinterfaces/mantidqtinterfaces/PyChop/PyChopGui.py_ . (Surprised we're even allowed to have PyChop.py _and_ a directory named PyChop. How does one import `PyChop/__init__.py`?) But the rest of the modules under PyChop don't deal with qt or GUIs - we could move most of this to `scripts` and avoid the problem.

But it looks like this code was moved _from_ scripts quite recently https://github.com/mantidproject/mantid/commit/5654ce0702485141383191770e416ea77d342591 so maybe there is a reason to prefer some other option?

ajjackson committed:
Move non-Qt parts of PyChop to scripts

This partially reverts a previous move of code from scripts to
mantidqtinterfaces. The reason is that Abins2D needs to access parts
of the pychop Python code that do not involve Qt, but this is leading
to test failures which require Qt code and scripts to be
well-separated.

As a possible solution, this commit moves _most_ of pychop into
scripts, but leaves PyChop.py and PyChopGui.py in their current
locations. There might be a tidier way of arranging this, or at least
naming things? For now we are using capitalisation to distinguish
PyChop (Qt-related) from pychop (scripts) which doesn't seem ideal.

ajjackson committed:
Abins2D: Converty Pychop resolution from FWHM to sigma

Abins/Pychop resolutions appear too broad compared to experiment, by a
factor of about 2. Presumably this is because we are
interpreting a full-width-half-maximum value as the Gaussian sigma;
these are related by a scale factor close to 2.35.

We should check how the resolution for PANTHER was defined; this is
likely to also be a FWHM but I haven't got this written explicitly.

ajjackson committed:
Abins2D: Treat PANTHER resolution function as FWHM

Awaiting confirmation, but this seems a more likely interpretation
than the current treatment (standard deviation).

ajjackson committed:
Abins2D: Drop bin_width and sample_form options

These are not very useful

- bin_width only accepts a narrow range of values anyway. For 2-D
  instruments the energy range is often much smaller than TOSCA, so
  smaller bins might be needed for a good plot. This should be
  controlled by advanced parameters instead of cluttering up the main
  algorithm; generally we would like to make a recommendation and
  users should not need to tweak it.

- sample_form still only has one option. Although ideas are developing
  for how AbINS might handle crystalline systems (now that force
  constants data can be imported), it is not clear that "Sample Form"
  is the user option that will map to this.

We should consider removing it from Abins (1D) as well, but for now
let's at least avoid needlessly introducing them in the release
version of Abins2D.

ajjackson committed:
Abins: overhaul bin_width parameter

So far Abins has relied on a singled bin_width parameter, modified at
runtime by the user input.

- For direct-geometry instruments the energy range is variable, so it
  makes more sense to specify a number of bins than width.

- For indirect-geometry the energy range is generally fixed and it is
  more natural to specify a width; but we may like to have a different
  default per instrument.

- abins.parameters.sampling['bin_width'] remains the most
  natural/convenient place for advanced users to tinker with this
  parameter, so we make this highest-priority parameter, unset by
  default.

- As instruments are already responsible for proving energy range, it
  is quite natural to move other energy bin information (bin width and
  actual bin set) to getters on the Instrument.

ajjackson committed:
Abins2D: Make energy limits check aware of chosen units

- Previously this was buggy, as the ref value in wavenumber would be
  compared with whatever number the user provides (which may be in
  meV!)

- As well as making the comparison correct, the invalid parameter
  tooltip is now also given in the user's chosen units. In practice this
  is *much* nicer to use.

ajjackson committed:
Abins: cache instrument setting data on getter

Instrument setting is immutable, so this only affects the case where
advanced parameters are changed after the Instrument was
instantiated - which Abins never does.

As well as saving some overhead, this removes the excessive logging
when using default setting. (Currently a message is logged every time
a parameter is checked with these methods...)

ajjackson committed:
Abins2D: fix DW application when no autoconvolution

This was an oversight leading to a "referenced before assignment"
error. The issue was already fixed in the Abins1D code, but in this
duplicated region the fix was overlooked. Most manual testing used
autoconvolution so it wasn't noticed for a while.

Anyway, this duplication needs to be cut down somehow.

ajjackson committed:
Abins: drop redundant bin_width parameter from calculator

Instrument is now aware of bin_width, and already included in args.

Testing revealed a discrepancy in the bin_width defaults
implementation; key name needed updating.

ajjackson committed:
Abins2D tidy-up: track 'isotropic_fundamentals' with property

This rarely-used argument is inconsistent with other parameters being
set on provate properties. Much as I'd like to make some of this code
more explicit, at the moment it's just adding clutter without moving
toward that goal.

ajjackson committed:
Refactoring abins powder calculator: bring together 1-D, 2-D

We can cut a lot of redundant code across the 1-D and 2-D
implementations, but it involves a bit of array-twiddling.

ajjackson committed:
Abins tweak: correct shape of iso_dw for 1-D / 2-D cases

Axis swap is only appropriate for 2-D case.

I don't understand how this wasn't breaking the tests.

ajjackson committed:
Abins: Expand 1d tests for main intensity calculator

Add separate tests for order-2 and order-10 (i.e. o2+autoconv)
calculations. There are program-flow differences in how these
calculations are performed so it is important to cover them.

ajjackson committed:
Abins 2D: extend CalculateSPowderTest to test 2d, higher orders

- Add test data for 2D instruments: one pychop-based case (Mari) and
  one special instrument (Panther). Both include NaN masking of
  kinematically-forbidden region.

- Advanced parameters are used to evaluate data on a much coarser mesh
  than usual, reducing reference file size to manageable ~100 KB.

- Testing of cache reload is disabled for 2D instruments, as this
  feature is currently disabled. (It costs time and is of little
  benefit for such calculations in its current form.)

- 2D data breaks assumption of data handlers that the number of atoms
  will equal number of fields - 1. Make this more robust.

- Existing tests only check fundamental spectrum, but other orders are
  still there in the test file? We really should be checking all the
  orders, it is quite possible an error will affect higher orders
  without changing fundamentals. Extended check to loop over orders.

ajjackson committed:
Abins: fix order-2 spectra regression from refactor

q^2n scaling was broken when not including autoconvolution terms

This was not noticed because
- Manual testing tended to have autoconvolution enabled
- Unit tests were running this step but not checking those
  terms (before the bug was introduced)

The problem was caught by system tests :D

Test data introduced after the bug has been updated.

ajjackson committed:
Abins2D: new docs page

- Use seeAlso attribute to refer between Abins algos
- A few working tweaks and cuts that might make their way back to 1-D
  Abins page (maybe package with cut-back input options?)
- Doctests are probably broken, still need to sort those out

ajjackson committed:
Abins2D: add release note

ajjackson committed:
Abins2D: doctest

ajjackson committed:
Abins2D: rename TwoDMap to Ideal2D

This makes it clearer that this is a special "non-instrument"
simulation. _All_ the 2D instruments are 2-D maps...

ajjackson committed:
Abins2D: add a pretty picture to release note

ajjackson committed:
Abins2D: simple systemtest

The systemtest setup didn't consider Abins2D. A bit of refactoring
could be done to handle this (and promote re-use of input files) but
for the sake of one test case we don't really need it yet.

ajjackson committed:
Abins2D: fix Ideal2D, non-chopper instruments

- Only sanity-check chopper parameters for chopper instruments. Should
be fine to leave these blank for Ideal2D or PANTHER
- This lets us clean up a little, set null choppers for Ideal2D and
  PANTHER and get nonsense G004 out of chopper list.
- The other doctest didn't have the right algorithm name! Easy fix,
  let's see if it's enough.

ajjackson commented:
Once again, I am pushing something that was _just_ rebased against main and is marked as having merge conflicts. I think something has gone wrong with that process.

AnthonyLim23 commented on Framework/PythonInterface/plugins/algorithms/Abins2D.py:
```diff
@@ -0,0 +1,291 @@
+        if self._save_ascii:
+            self.write_workspaces_to_ascii(ws_name=self._out_ws_name,
+                                           scale=(1.0 / self._instrument.get_energy_bin_width()))
+            prog_reporter.report("All workspaces have been saved to ASCII files.")
```
is it worth reporting the path to the files?

ajjackson replied:
That's not existing behaviour but does seem helpful. Maybe a new Issue?

ajjackson replied:
Issue created https://github.com/mantidproject/mantid/issues/33993

AnthonyLim23 commented on Framework/PythonInterface/plugins/algorithms/Abins2D.py:
```diff
@@ -0,0 +1,291 @@
+
+        if self._instrument.get_name() not in TWO_DIMENSIONAL_INSTRUMENTS:
+            raise ValueError("Instrument {self._instrument_name} is not supported by this version of Abins")
+
```
```suggestion
elif protons_number > nucleons_number:
    raise ValueError("The proton number is greater than the nucleon number")
```
or something similar

ajjackson replied:
If this is going to be checked, it would make more sense to put it in the AtomsData class so it is caught as soon as the data is loaded. I think this can be a separate issue?

ajjackson replied:
Issue created
https://github.com/mantidproject/mantid/issues/34004

AnthonyLim23 commented on Testing/SystemTests/tests/framework/AbinsTest.py:
```diff
@@ -485,3 +485,44 @@ def runTest(self):
+                ScaleByCrossSection=self._cross_section_factor, OutputWorkspace=self._output_name,
+                **params_2d)
+
+    def validate(self):
```
whats this used for?

ajjackson replied:
It's a standard SystemTest, works the same way as the others in this file. https://developer.mantidproject.org/SystemTests.html#writing-a-test

HelperTestingClass sets `self._output_name`, the test framework calls `runTest()` for the calc and compares the workspaces returned by `validate()`.

AnthonyLim23 commented on docs/source/algorithms/Abins2D-v1.rst:
```diff
@@ -0,0 +1,107 @@
+.. algorithm::
```
up to here

AnthonyLim23 commented on docs/source/algorithms/Abins2D-v1.rst:
```diff
@@ -0,0 +1,107 @@
+
+Abins2D is a plugin for Mantid which allows scientists to generate theoretical inelastic neutron scattering spectra (INS) in 2-D :math:`(|\mathbf{q}|,\omega)` space, accounting for the limitations of real-world measurements.
+
+The intensity is calculated in the "almost-isotropic" incoherent powder-averaging approximation, as used in the 1-D Abins algorithm.
```
lets make the 1-D Abins algorithm a link to the doc

ajjackson replied:
Do you think
```
ref:`Abins`
```

will work?

AnthonyLim23 replied:
I think you need to do it the same as in the release notes

ajjackson replied:
```
ref:`Abins <algm-Abins>`
```
does the trick in main docs. I can't tell if it works for the release not because I don't remember the incantation to rebuild the release notes cache without rebuilding all of Mantid as well :sob:

AnthonyLim23 commented on docs/source/algorithms/Abins2D-v1.rst:
```diff
@@ -0,0 +1,107 @@
+Different 2-D INS instruments may be selected, along with instrument parameters (incident energy, chopper settings) that determine the accessible :math:`(|\mathbf{q}|,\omega)` region and energy resolution.
+More information about the implemented working equations can be found :ref:`here <DynamicalStructureFactorFromAbInitio>`.
+
+Abins requires data from existing lattice dynamics calculations to determine the phonon mode frequencies and displacements. These are used to compute INS intensities; higher-order phonons may approximated as combinations of fundamental modes.
```
```suggestion
Abins requires data from existing lattice dynamics calculations to determine the phonon mode frequencies and displacements. These are used to compute INS intensities; higher-order phonons may be approximated as combinations of fundamental modes.
```

AnthonyLim23 commented on docs/source/algorithms/Abins2D-v1.rst:
```diff
@@ -0,0 +1,107 @@
+Abins requires data from existing lattice dynamics calculations to determine the phonon mode frequencies and displacements. These are used to compute INS intensities; higher-order phonons may approximated as combinations of fundamental modes.
+Several file formats are supported for frequency and displacement data at given :math:`\mathbf{q}`-points: CASTEP (.phonon), CRYSTAL (.out), GAUSSIAN (.log), DMOL3 (.outmol) or VASP (.xml).
+For VASP XML inputs, it is also permitted to use a "selective dynamics" in which some atoms are frozen.
+These would typically form a rigid substrate (e.g. a noble metal surface) for an adsorbed molecule of lightweight atoms, and are omitted from the calculated spectrum.
```
Is this only true for the selective dynamics?

ajjackson replied:
It is really, rigid surfaces absolutely do vibrate. The approximation only works if you're able to neglect the intensity of the substrate vibrations, or they are active in a different frequency range from the adsorbate. Both of which are fair assumptions if you are looking at stretching modes of hydrogenous molecules.

AnthonyLim23 replied:
could we make it clearer in the docs? I was unsure when reading it

ajjackson replied:
Abins will accept your frozen atoms regardless of whether they are a good physical approximation.

I guess I can be more explicit that you probably shouldn't do this in other cases?

AnthonyLim23 commented on docs/source/release/v6.4.0/Direct_Geometry/Algorithms/New_features/33390.rst:
```diff
@@ -0,0 +1,10 @@
+- A new Abins2D algorithm has been added for simulation of inelastic neutron scattering in direct-geometry
+
+  - This employs the same almost-isotropic analytic powder-averaging incoherent approximation as Abins
```
link to abins doc

AnthonyLim23 commented on docs/source/release/v6.4.0/Direct_Geometry/Algorithms/New_features/33390.rst:
```diff
@@ -0,0 +1,10 @@
+
+  - This employs the same almost-isotropic analytic powder-averaging incoherent approximation as Abins
+  - The ISIS MARI, MAPS, and MERLIN instruments are supported, alongside ILL-PANTHER and a "TwoDMap" ideal simulation. Please direct requests/specifications for additional instruments to Sanghamitra Mukhopadhyay and Adam Jackson.
+  - The algorithm produces workspaces in (E,q) space, best viewed interactively in Sliceviewer, or rendered as colorfill plots
```
do we want the E and q to be in "math font"?

ajjackson replied:
Then I'd have to decide what to typeset them consistently with :laughing: . (*Q,ω*) probably best...

AnthonyLim23 commented on scripts/abins/instruments/broadening.py:
```diff
@@ -528,13 +528,13 @@ def interpolated_broadening(sigma=None, points=None, bins=None,
-    # Sample on appropriate log scale: log_b(x) = log(x) / lob(b)
-    n_kernels = int(np.ceil(np.log(max(sigma) / min(sigma)) / np.log(log_base)))
+    # Sample on appropriate log scale: log_b(x) = log(x) / log(b)
+    n_kernels = int(np.ceil(np.log(max(sigma) / min(sigma)) / np.log(log_base))) + 1
```
whats the plus 1 for?

ajjackson replied:
Ensures there is always a higher kernel available so interpolation has two end groups to work between. The existing setup breaks if you try to interpolate to a width that is exactly equal to the highest width; in practice this happens for constant-width broadening (where only one kernel is really needed) and this PR introduces the "ideal simulation" 2-D instrument which has such a fixed width.

In principle it would be more efficient to
- fall back on non-interpolated broadening for fixed-width cases
- otherwise, only add the extra point when the rare but theoretically-possible max(sigma)==min(sigma)*factor^n_kernels case arises.

But the +1 is a simple fix and not a big overhead.

AnthonyLim23 commented on scripts/abins/instruments/directinstrument.py:
```diff
@@ -0,0 +1,156 @@
+from __future__ import (absolute_import, division, print_function)
```
I dont think we need this line anymore .

ajjackson replied:
Python is dead, long live Python

AnthonyLim23 commented on scripts/abins/instruments/directinstrument.py:
```diff
@@ -0,0 +1,156 @@
+            min_angle, max_angle = -max_angle, -min_angle
+        elif (min_angle < 0):
+            # e.g. if -10 < 2θ < 20, we don't want to set limits 10 < |2θ| < 20
+            # as this would omit contributions from the range -10 < θ < 10
```
nice comment

ajjackson replied:
It took a while to figure out all these cases :sweat_smile:

AnthonyLim23 commented on scripts/abins/instruments/directinstrument.py:
```diff
@@ -0,0 +1,156 @@
+        k2_i[conservation_indx] = e_init * WAVENUMBER_TO_INVERSE_A
+
+        cos_angle = math.cos(math.radians(angle))
+        result = k2_i + k2_f - 2 * cos_angle * (k2_i * k2_f) ** 0.5
```
```suggestion
        return k2_i + k2_f - 2 * cos_angle * np.sqrt(k2_i * k2_f)
```

AnthonyLim23 commented on scripts/abins/instruments/directinstrument.py:
```diff
@@ -0,0 +1,156 @@
+
+        if scheme in prebin_required_schemes:
+            s_dft, _ = np.histogram(frequencies, bins=bins, weights=s_dft, density=False)
+            frequencies = (bins[1:] + bins[:-1]) / 2
```
Are these the frequencies as point data?

ajjackson replied:
They are frequencies corresponding to the weights. This can be bin centres of pre-binned data, or a full set of frequencies and weights for individual modes. (Which might be automatically binned here if necessary.)

AnthonyLim23 replied:
what would be better name? frequencies_of_weights?

ajjackson replied:
`frequencies` is used pretty consistently around Abins for this data, I think calling it anything else here would be more confusing.

ajjackson replied:
Perhaps it would help to complete this docstring with more information about the _output_, clarifying that it will always be binned?

ajjackson replied:
How about:
- points_freq renamed to bin_centres
- docstring becomes
```
        """
        Convolve discrete frequency spectrum with energy-dependent resolution function for a 2D instrument.

        :param frequencies: DFT frequencies for which resolution function should be calculated (frequencies in cm-1)
        :param s_dft:  discrete S calculated directly from DFT

        :returns:
            (bin_centres, broadened_spectrum):
            Resolution functions always create binned data on a regular mesh.
            ``bin_centres`` and ``broadened_spectrum`` are a consistent set of
            mesh frequencies and corresponding broadened intensity.
        """

AnthonyLim23 commented on scripts/abins/instruments/instrument.py:
```diff
@@ -19,14 +23,104 @@ class Instrument:
+            default: Value to be returned if key is not found.
+        """
+        setting_parameters = self.get_setting_parameters()
+        if key in setting_parameters:
```
is this supposed to give priority to the value in settings?

ajjackson replied:
Yes, settings can overrule non-setting-specific parameters. I'm not sure it makes any difference in practice as they are (so far) never specified in both places; it's just an efficient way of checking both places and skipping the second check if something was already found.

It makes more sense for the setting parameters to overrule the generic ones than vice versa if the situation _does_ arise. (Say one instrument has a whole bunch of settings, and only one modifies the resolution parameters; then the other settings could fall back to an instrument-wide value.)

AnthonyLim23 commented on scripts/abins/instruments/instrument.py:
```diff
@@ -19,14 +23,104 @@ class Instrument:
+        """
+        bin_width = abins.parameters.sampling['bin_width']
+        if not bin_width:
+            bin_width = self.get_parameter('energy_bin_width', default=None)
```
will this and line 107 always return None if there is no value?

ajjackson replied:
`abins.parameters.sampling['bin_width']` is
- defined as None in the parameters file file, but
- actually a slightly special case; it is used by the Abins Algorithm to override other bin width settings. The default value for BinWidthInWavenumber (which will be passed through) is 1.0

The user is not allowed to use the Abins Algorithm to set bin_width to 0.0; validateInputs requires it to be between 1.0 and 10.0.

(I think this is overly restrictive, and also not an important option to offer through the Algorithm/GUI. I'd like to ultimately remove it and use the same scheme as Abins2D: Instruments have defaults, and if you want to set a custom value you need to mess around with abins.parameters. Maybe then we can provide some kind of sensible dictionary **kwargs option for advanced parameters to be passed around in a less hack-y way.)

AnthonyLim23 commented on scripts/abins/instruments/pychop.py:
```diff
@@ -0,0 +1,80 @@
+from __future__ import (absolute_import, division, print_function)
```
I dont think this line is needed

AnthonyLim23 commented on scripts/abins/instruments/pychop.py:
```diff
@@ -0,0 +1,80 @@
+
+        self._pychop_instrument = pychop_instruments.Instrument(
+                self._name, chopper=setting, freq=chopper_frequency)
+        self._tthlims = self._pychop_instrument.detector.tthlims
```
what is tthlims?

AnthonyLim23 replied:
yeah that would be good. I would have never guessed it was two theta

AnthonyLim23 commented on scripts/abins/sdata.py:
```diff
@@ -152,6 +183,29 @@ def get_empty(cls: SD, *,
+        """Update q-bins stored on SData
+
+        Args:
+            q_bins: 1-D set of q-point bin edges surrounding rows in S data
```
can we have an example? It might make it easier to understand

ajjackson replied:
Would
```
(e.g. for q-bins=[1., 2., 3.], S arrays will have two rows
corresponding to q=1-2 Å^-1, q=2-3 Å^-1; typically these
are evaluated at the bin centres 1.5, 2.5 Å^-1.)
```
do the trick?

AnthonyLim23 replied:
swap the dash for a to and yes (since we are using dash for minus in the same statement)

AnthonyLim23 commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -127,6 +128,16 @@ def __init__(self, *,
             self._fine_bins = None
             self._fine_bin_centres = None
 
+        # If operating in 2-D mode, there is also an explicit set of q bins
```
up to here

AnthonyLim23 commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -353,33 +384,103 @@ def _calculate_s_powder_over_k(self, *, angle: float,
+            SData
+        """
+        # Initialize the main data container
+        # sdata = self._get_empty_sdata(use_fine_bins=self._autoconvolution,
```
is this supposed to be here still?

AnthonyLim23 commented on scripts/pychop/__init__.py:
```diff
@@ -0,0 +1,26 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
```
up to here

AnthonyLim23 commented on scripts/test/Abins/AbinsCalculateSPowderTest.py:
```diff
@@ -123,30 +188,33 @@ def _prepare_data(self, filename=None):
 
-            temp = np.asarray(correct_data["atom_%s" % el]["s"]["order_%s" % FUNDAMENTALS])
-            correct_data["atom_%s" % el]["s"]["order_%s" % FUNDAMENTALS] = temp
+            # correct_data[f"atom_{i}"]["s"]["order_1"] = np.asarray(
```
are these meant to be here still?

ajjackson commented on Framework/PythonInterface/plugins/algorithms/Abins.py:
```diff
@@ -19,42 +18,41 @@
-        self.declareProperty(name="Autoconvolution", defaultValue=False,
-                             doc="Estimate higher quantum orders by convolution with fundamental spectrum.")
+        # Soon-to-be-deprecated properties (i.e. already removed from 2D)
+        self.declareProperty(name="BinWidthInWavenumber", defaultValue=1.0,
```
No, currently this will be called every time the Algorithm is run. Deprecation warnings are usually used to recommend users adopt the future behaviour, which won't be possible yet.

ajjackson commented on Framework/PythonInterface/plugins/algorithms/Abins2D.py:
```diff
@@ -0,0 +1,291 @@
+            partial_wrk_names = []
+
+            for n in range(dim):
+                seed = "quantum_event_%s" % (n + 1)
```
Probably inherited from the CASTEP convention where files are referred to as e.g. "seedname.param". It's a bad name. For consistency, it should also change in Abins.py

Actually, this doesn't need to be a variable at all. We have f-strings now so it's quite clean to build the string as a one-liner.

ajjackson commented on Framework/PythonInterface/plugins/algorithms/Abins2D.py:
```diff
@@ -0,0 +1,291 @@
+
+        freq_axis = NumericAxis.create(n_freq_bins)
+
+        # q_size = abins.parameters.instruments[self._instrument.get_name()]['q_size']
```
Nope, it's cleaner to use self._q_bins. Will delete.

ajjackson committed:
Apply suggestions from code review

Co-authored-by: Anthony <anthony.lim@stfc.ac.uk>

ajjackson committed:
Abins2D tweaks from reviewer comments

Thanks @AnthonyLim23

ajjackson commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -353,33 +384,103 @@ def _calculate_s_powder_over_k(self, *, angle: float,
+            iso_dw = self.calculate_isotropic_dw(q2=q2[order_expansion_slice[:-1]])
+
+            if len(iso_dw.shape) == 3:
+                # From q, atom, freq -> atom, q, freq
```
We need to get a consistent indexing convention at this point so that the same `sdata.apply_dw()` works for 1-D and 2-D data.

The steps to calculate isotropic_dw use quite a bit of numpy broadcasting. In the 2-D case `q2` is a column vector (creating the 2-D map when combined with a row of energy data) while in 1-D case `q2` is a row vector corresponding to energy.
Perhaps it would make more sense to include this axis check and swap inside `calculate_isotropic_dw`? It might be possible to craft the data input to `_isotropic_dw` more carefully to avoid the axis swap, but I doubt it would reduce complexity.

AnthonyLim23 replied:
it just seemed odd without the context. A check on the axis would probably be good

AnthonyLim23 commented on scripts/abins/sdata.py:
```diff
@@ -152,6 +183,32 @@ def get_empty(cls: SD, *,
+        Args:
+            q_bins: 1-D set of q-point bin edges surrounding rows in S data
+                (e.g. for q-bins=[1., 2., 3.], S arrays will have two rows
+                 corresponding to q=1-2 Å^-1, q=2-3 Å^-1; typically these
```
```suggestion
                 corresponding to q =1 to 2 Å^-1, q =2 to 3 Å^-1; typically these
```
just so no-one tries to read it as 1 minus 2

ajjackson committed:
Clarify docstring in scripts/abins/sdata.py

Co-authored-by: Anthony <anthony.lim@stfc.ac.uk>

ajjackson committed:
Abins clarity improvements: frozen-atom methods, tthlims variable

ajjackson committed:
Abins2D: fix sphinx links to algm-Abins

Seems to work for Abins2D page. I can't remember how to rebuild the
release notes, some kind of cache needs clearing? But it should work
there too.

ajjackson committed:
Abins: clear up some internal notation for broadened data

Requested by @AnthonyLim23

Make it a bit clearer that the output mesh is always rebinned, whereas
input frequencies can be either or rebinned on-the-fly.

ajjackson committed:
Abins: small refactor based on code review

There's a bit of book-keeping of array indices needed to use numpy
broadcasting and handle both 1-D and 2-D without too much extra
code. Move an axis-swapping tidy-up operation into the method that
creates the data, rather than a higher-level control-flow function.

AnthonyLim23 commented:
also need to test pychop:

- [ ] pychop GUI
- [ ] crystalField @SilkeSchomann or @MialLewis could you please test this?

AnthonyLim23 commented:
the abins stuff looks good

AnthonyLim23 commented:
pychop seems to work as far as I can tell. I would like it if someone else could also have a look

SilkeSchomann commented:
CrystalField looks good as well

ajjackson commented:
> ModuleNotFoundError: No module named 'pychop'
> 
> Traceback (most recent call last): File "c:\users\eoc57742\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 402, in lambda checked_py, script=script: self.launch_custom_python_gui(script)) File "c:\users\eoc57742\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 360, in launch_custom_python_gui self.interface_executor.execute(open(filename).read(), filename) File "c:\users\eoc57742\mantid\qt\python\mantidqt\mantidqt\widgets\codeeditor\execution.py", line 150, in execute executor.execute(code_str, filename, flags, self.globals_ns, line_offset) File "c:\users\eoc57742\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\PyChop.py", line 14, in from mantidqtinterfaces.PyChop import PyChopGui File "c:\users\eoc57742\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\PyChop__init__.py", line 26, in from pychop.Instruments import Instrument as PyChop2 # noqa: F401 ModuleNotFoundError: No module named 'pychop'

Oh dear! I can't reproduce this, what build/system is it? How did you try to open pychop?

(I remain deeply suspicious of the fact that mantidqtinterfaces has both a PyChop.py and a Pychop/\_\_init\_\_.py, but that doesn't seem to be the problem here.)

ajjackson commented:
I'm going to try building a Mac package and opening the PyChop interface in that.

SilkeSchomann commented:
Debug build on Windows in my case. No luck even with a clean build.

AnthonyLim23 commented:
@SilkeSchomann are you using conda?

ajjackson commented:
Does `pychop.Instruments import Instrument as PyChop2` work from the Mantid Python interpreter?

SilkeSchomann commented:
> 

No, it doesn't:

ModuleNotFoundError: No module named 'pychop'
  File "<string>", line 6, in <module>

ajjackson commented:
Is there a `pychop` directory in your scripts folder?

Wondering if there's some build file that needs to be updated...

SilkeSchomann commented:
The existing PyChop folder is not in lower case so I guess when merging the PR code in it does not change it. However, when manually changing it to all lower case I can open the PyChop interface.

SilkeSchomann commented:
The PyChop interface seems to work fine. Would it be possible to change pychop to PyChop? I think otherwise it might cause problems for other users.

ajjackson commented:
I'm not sure that will work for the Qt Interface code, which is also called PyChop, and needs to be able to find both. The casing allows us to differentiate between them (and adopt a more pythonic library name...)

Maybe the part in _scripts_ should be called something like `pychoplib`?

ajjackson commented:
I've just tried the Mac package and it seems fine. I think this might only happen for developer builds from Git where there is an existing PyChop folder? (i.e. from before it was moved the first time). Git is sometimes funny about leaving folders behind because it really tracks files rather than directories. Then when creating the "new" _pychop_ there might have been a Windows case-insensitivity issue...

Is your build a local build, or have you tried a Windows package? I suspect the "clean build" may not have been from clean _source_.

SilkeSchomann commented:
It works since I renamed the pychop folder in scripts to all lower case. I can both open the PyChop interface and also use from pychop.Instruments import Instrument as PyChop2

I had cleaned the Windows build in Visual Studio and run cmake. After compiling everything (and then renaming the pychop folder) it started to work.

I've used test-pr to merge the PR code and did not check out the branch for the PR directly though.

SilkeSchomann commented:
It is possible that this only is an issue for a developer build and not the release. Maybe we should re-test this once it is in nightly.

AnthonyLim23 commented:
I think as Adam has said it could be left over folders from previous versions. One way to "check" is to just force delete everything in the offending parent folder (e.g. scripts). Then do a `git reset --hard` and it should "restore" the actual files in scripts. So any old folders will be gone.

ajjackson commented:
I suspect it is only a problem for developer builds that are in a particular "contaminated" state ;-) The main branch does not have a scripts/PyChop so I think the merge on Github will be fine.

AnthonyLim23 reviewed (approved):
Works as described. It doesn't move any qt and from testing pychop is unaffected by the moving of the files.
